File: check_dig/check_dig.py
#!/usr/bin/env python3

"""
MIT License

Copyright (c) 2023-2025 Thomas Vincent

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

This script performs DNS queries via SSH.

It connects to a remote SSH host, executes a DNS query using the `dig` command,
and returns the results. It includes robust retry mechanisms, timeout handling,
and allows specification of warning and critical thresholds for monitoring.

Usage:
  dns_query.py -R <SSH_HOST> -l <QUERY_ADDRESS> [-P <SSH_PASSWORD>] 
               [-U <SSH_USERNAME>] [-s <SSH_PORT>] [-p <DNS_PORT>] 
               [-T <RECORD_TYPE>] [-a <EXPECTED_ADDRESS>] 
               [-A <DIG_ARGUMENTS>] [-w <WARNING>] [-c <CRITICAL>] 
               [-t <TIMEOUT>] [-r <RETRIES>]

Example:
  dns_query.py -R example.com -l www.google.com -T A
"""

import argparse
import subprocess
import sys
import time

# SSHConnection class handles the execution of SSH commands on remote hosts.
class SSHConnection:
    def __init__(self, hostname, username=None, password=None, port=22):
        self.hostname = hostname
        self.username = username
        self.password = password
        self.port = port

    def execute_command(self, command, timeout):
        """Executes a command on the remote SSH host with a specified timeout."""
        ssh_cmd = ["ssh", "-p", str(self.port)]
        if self.username:
            ssh_cmd.extend(["-l", self.username])
        ssh_cmd.append(self.hostname)
        ssh_cmd.append(command)

        result = subprocess.run(
            ssh_cmd,
            capture_output=True,
            text=True,
            timeout=timeout
        )
        if result.returncode != 0:
            raise subprocess.CalledProcessError(
                result.returncode, ssh_cmd, result.stdout, result.stderr)

        return result.stdout.strip()

# DNSQuery class performs DNS queries via SSH, manages retries, and checks thresholds.
class DNSQuery:
    def __init__(self, ssh_connection, query_address, dns_port=53, record_type='A',
                 dig_arguments=None, expected_address=None, timeout=10, retries=1,
                 warning=None, critical=None):
        self.ssh_connection = ssh_connection
        self.query_address = query_address
        self.dns_port = dns_port
        self.record_type = record_type
        self.dig_arguments = dig_arguments
        self.expected_address = expected_address
        self.timeout = timeout
        self.retries = retries
        self.warning = warning
        self.critical = critical

    def run(self):
        """Runs DNS query with retries and handles response checking."""
        attempt = 0
        while attempt < self.retries:
            try:
                start_time = time.time()
                response = self.query_dns()
                duration = time.time() - start_time
                self.handle_result(response, duration)
                return
            except (subprocess.TimeoutExpired, subprocess.CalledProcessError) as e:
                print(f"Attempt {attempt + 1} failed: {e}")
                attempt += 1
                if attempt < self.retries:
                    print(f"Retrying in {self.timeout} seconds...")
                    time.sleep(self.timeout)
                else:
                    sys.exit(2)

    def query_dns(self):
        """Executes the DNS query via SSH using `dig`."""
        dig_cmd = f"dig -p {self.dns_port} -t {self.record_type} +short {self.query_address}"
        if self.dig_arguments:
            dig_cmd += f" {self.dig_arguments}"
        return self.ssh_connection.execute_command(dig_cmd, timeout=self.timeout)

    def handle_result(self, response, duration):
        """Evaluates DNS query response against expected results and timing thresholds."""
        if self.expected_address and self.expected_address not in response:
            print(f"CRITICAL: Expected address '{self.expected_address}' not found. Response: {response}")
            sys.exit(2)

        if self.critical and duration > self.critical:
            print(f"CRITICAL: Query took {duration:.2f}s")
            sys.exit(2)
        elif self.warning and duration > self.warning:
            print(f"WARNING: Query took {duration:.2f}s")
            sys.exit(1)

        print(f"OK: DNS response '{response}' received in {duration:.2f}s")
        sys.exit(0)

# parse_args function parses command-line arguments for the script.
def parse_args():
    parser = argparse.ArgumentParser(description="DNS query via SSH")

    parser.add_argument("-R", required=True, help="SSH hostname or IP")
    parser.add_argument("-l", required=True, help="DNS query hostname or IP")
    parser.add_argument("-U", help="SSH username")
    parser.add_argument("-P", help="SSH password")
    parser.add_argument("-s", type=int, default=22, help="SSH port (default: 22)")
    parser.add_argument("-p", type=int, default=53, help="DNS port (default: 53)")
    parser.add_argument("-T", default="A", help="DNS record type (default: A)")
    parser.add_argument("-A", help="Additional dig arguments")
    parser.add_argument("-a", help="Expected DNS response")
    parser.add_argument("-w", type=float, help="Warning threshold in seconds")
    parser.add_argument("-c", type=float, help="Critical threshold in seconds")
    parser.add_argument("-t", type=int, default=10, help="Timeout in seconds (default: 10)")
    parser.add_argument("-r", type=int, default=1, help="Number of retries (default: 1)")

    return parser.parse_args()

# main function sets up and executes the DNS query.
def main():
    args = parse_args()

    ssh_conn = SSHConnection(args.R, args.U, args.P, args.s)
    dns_query = DNSQuery(
        ssh_conn,
        args.l,
        dns_port=args.p,
        record_type=args.T,
        dig_arguments=args.A,
        expected_address=args.a,
        timeout=args.t,
        retries=args.r,
        warning=args.w,
        critical=args.c
    )

    dns_query.run()

if __name__ == "__main__":
    main()

File: check_dig/check_dig.c
/*
 * dns_lookup.c
 *
 * This program uses the dig command to lookup a DNS record.
 *
 * Copyright (c) 2023-2024 Thomas Vincent
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
 * Usage:
 *
 *   dns_lookup <query_address> [-p <server_port>] [-t <record_type>]
 *
 * Options:
 *
 *   -p <server_port>: The port number of the DNS server.
 *   -t <record_type>: The type of DNS record to lookup.
 *
 * Example:
 *
 *   dns_lookup www.google.com
 *
 * This will lookup the A record for www.google.com.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define MAX_COMMAND_LENGTH 1024

// Function to print errors to stderr
void print_error(const char *message) {
  fprintf(stderr, "ERROR: %s\n", message);
}

// Function to print the command being executed (for debugging)
void print_command(const char *command) {
  printf("Running command: %s\n", command);
}

// Function to execute a command using system()
int run_command(const char *command) {
  return system(command);
}

// Function to read the output of a command (from stdout)
// Note: This function is not used in the current code and can be removed
char *read_command_output() {
  char *output = NULL;
  size_t output_size = 0;
  ssize_t bytes_read = getline(&output, &output_size, stdin);
  if (bytes_read < 0) {
    print_error("Could not read output of command");
    free(output);
    return NULL;
  }
  return output;
}

// Function to parse the output of the dig command
// This function currently just prints the record and address. 
// It can be extended to return specific information or perform more advanced parsing.
char *parse_output(const char *output, const char *expected_address) {
  char *record = NULL;
  char *address = NULL;
  char *colon = strchr(output, ':');
  if (colon == NULL) {
    print_error("Could not find colon in output of command");
    return NULL;
  }
  record = strndup(output, colon - output);
  address = strndup(colon + 1, strlen(output) - (colon - output) - 2); // -2 to remove newline and space

  if (expected_address != NULL && strcmp(expected_address, address) != 0) {
    print_error("Expected address not found");
    free(record);
    free(address);
    return NULL;
  }

  printf("DNS %s - %s\n", record, address);
  free(record);
  free(address);

  return (char *)output; // Cast to avoid warning
}

int main(int argc, char *argv[]) {
  // Get the command line arguments.
  char *query_address = NULL;
  int server_port = 53;
  char *record_type = "A";

  for (int i = 1; i < argc; i++) {
    if (strcmp(argv[i], "-p") == 0) {
      if (i + 1 < argc) {
        server_port = atoi(argv[i + 1]);
        i++;
      } else {
        print_error("No server port specified");
        return 1;
      }
    } else if (strcmp(argv[i], "-t") == 0) {
      if (i + 1 < argc) {
        record_type = argv[i + 1];
        i++;
      } else {
        print_error("No record type specified");
        return 1;
      }
    } else {
      if (query_address == NULL) {
        query_address = argv[i];
      } else {
        print_error("Too many arguments");
        return 1;
      }
    }
  }

  // Check if the query address is specified.
  if (query_address == NULL) {
    print_error("No query address specified");
    return 1;
  }

  // Run the command.
  char command_line[MAX_COMMAND_LENGTH];
  snprintf(command_line, sizeof(command_line), "dig @%s -p %d %s +short", 
           query_address, server_port, record_type);
  print_command(command_line);
  int status = run_command(command_line);

  // Check the status of the command.
  if (status != 0) {
    fprintf(stderr, "ERROR: Command failed with status %d\n", status);
    return 1;
  }

  // Parse the output of the command.
  // Since we're using `+short`, dig will output only the IP address
  char output[MAX_COMMAND_LENGTH];
  fgets(output, sizeof(output), stdin); 
  parse_output(output, NULL); // You can add expected address checking here

  return 0; 
}
